File: code/typescript/iterator/PostIterator.ts
import { Iterator } from "./Iterator";
import { Post } from "./Post";

export class PostIterator implements Iterator {
    private position = 0;

    constructor(private posts: Post[]) {}

    hasNext(): boolean {
        return this.position < this.posts.length;
    }

    next(): Post {
        if (!this.hasNext()) {
            throw new Error("No more posts available");
        }
        return this.posts[this.position++];
    }
}

File: code/python/iterator/PostIterator.py
from Iterator import Iterator


class PostIterator(Iterator):
    def __init__(self, posts):
        self.posts = posts
        self.index = 0

    def has_next(self):
        return self.index < len(self.posts)

    def next(self):
        if self.has_next():
            post = self.posts[self.index]
            self.index += 1
            return post
        else:
            raise StopIteration
